[shopsys] Elastic search filtering (#943)
shopsys/shopsys: docs/upgrade/UPGRADE-unreleased.md

    **Be careful, this upgrade will remove sessions**

### Application

>  - start filter products with elasticsearch ([#943](https://github.com/shopsys/shopsys/pull/943))

Reviewer: ```suggestion
 - start filtering products with elasticsearch ([#943](https://github.com/shopsys/shopsys/pull/943))
```

    - copy enhanced structure files from [GitHub](https://github.com/shopsys/project-base/blob/master/src/Shopsys/ShopBundle/Resources/definition-with-filter/products/) to `src/Shopsys/ShopBundle/Resources/definition-with-filter/products/` folder in your project (create `definition-with-filter` folder)

>     - set elasticsearch structure files directory in `app/config/paths.yml` to `'%shopsys.resources_dir%/definition-with-filter/'`

Reviewer: I'm missing an information about what variable should i set it to

    - replace class `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacade` with the interface `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface` in
        - `src/Shopsys/ShopBundle/Controller/Front/ProductController.php`
        - `src/Shopsys/ShopBundle/Controller/Front/SearchController.php`
    - add service definition for facade interface to your `services.yml` like `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface: '@Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainElasticFacade'`
    - add following alias to `services.yml` and `services_test.yml`
        ```yaml
        Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportRepository:

>             alias: Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportWithStructureRepository

Reviewer: ```suggestion
            alias: Shopsys\FrameworkBundle\Model\Product\Search\Export\ProductSearchExportWithFilterRepository
```

        ```

>     - fix test of export repository `tests/ShopBundle/Functional/Model/Product`

Reviewer: I would appreciate link with how should the file look like and name/class of the test

    - in order to have your code properly tested, copy following tests from [GitHub](https://github.com/shopsys/project-base/blob/master/tests/ShopBundle/Functional/Model/Product) to your project
        - `ProductOnCurrentDomainFacadeTest.php`
        - `ProductOnCurrentDomainElasticFacadeTest.php`
        - `ProductOnCurrentDomainSqlFacadeTest.php`
        - `ProductOnCurrentDomainFacadeCountDataTest.php`
        - `ProductOnCurrentDomainElasticFacadeCountDataTest.php`
        - `ProductOnCurrentDomainSqlFacadeCountDataTest.php`
        - `Filter/BrandFilterChoiceRepositoryTest.php`
        - `Filter/FlagFilterChoiceRepositoryTest.php`
        - `Filter/ParameterFilterChoiceRepositoryTest.php`
        - `Search/FilterQueryTest.php`
        - `Search/ProductSearchExportRepositoryTest.php`

>     - skip path `'*/tests/ShopBundle/Functional/Model/Product/ProductOnCurrentDomainFacadeCountDataTest.php'` for following coding standard sniffs

Reviewer: Where should I skip it (in what file)?

    - read the section about proxying the URL content subpaths via webserver domain [`docs/introduction/abstract-filesystem.md`](https://github.com/shopsys/shopsys/blob/master/docs/introduction/abstract-filesystem.md)
- to be more descriptive about error caused by active TEST environment ([#701](https://github.com/shopsys/shopsys/pull/701))
    - modify `ErrorController::createUnableToResolveDomainResponse()` by these [changes](https://github.com/shopsys/shopsys/pull/701/files#diff-0b1aecbf82624ce474ca3cb8bd75811c).

> - start filtering products with elasticsearch ([#943](https://github.com/shopsys/shopsys/pull/943))

Reviewer: We need split the upgrades to two sections - one for people that want to keep the SQl filter and the other one for the elastic filter

>     - replace Elasticsearch structure files in `src/Shopsys/ShopBundle/Resources/definition/products/` folder in your project with new ones from [GitHub](https://github.com/shopsys/project-base/blob/master/src/Shopsys/ShopBundle/Resources/definition/products/)

Reviewer: the link for GitHub is not working
```suggestion
    - replace Elasticsearch structure files in `src/Shopsys/ShopBundle/Resources/definition/products/` folder in your project with new ones from [GitHub](https://github.com/shopsys/project-base/tree/master/src/Shopsys/ShopBundle/Resources/definition/product/)
```

Author: link is to project-base repository as we agreed it will be better. Link will work after split.

    - replace class `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacade` with the interface `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface` in
        - `src/Shopsys/ShopBundle/Controller/Front/ProductController.php`
        - `src/Shopsys/ShopBundle/Controller/Front/SearchController.php`

>     - add service definition for facade interface to your `services.yml` like `Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface: '@Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainElasticFacade'`

Reviewer: This seems little more readable to me (please check if the formatting is good :smile:)
```suggestion
    - add service definition for facade interface to your `services.yml` like 
        ```yaml
        Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainFacadeInterface: '@Shopsys\FrameworkBundle\Model\Product\ProductOnCurrentDomainElasticFacade'
        ```
```
